src/frontend/context/ConfigContext.tsx

```diff
@@ -95,81 +95,87 @@ const defaultContext: ConfigContextType = [
 
 const ConfigContext = React.createContext<ConfigContextType>(defaultContext);
 
-type Props = {
-  children: React.ReactNode;
-};
-
-export const ConfigProvider = ({ children }: Props) => {
+export const ConfigProvider = ({ children }: React.PropsWithChildren<{}>) => {
+  const mountedRef = React.useRef(false);
   const [config, setConfig] = React.useState(defaultConfig);
   const [status, setStatus] = React.useState<Status>("idle");
-  const [reloadToken, setReloadToken] = React.useState();
   const intl = useIntl();
 
-  const reload = React.useCallback(() => setReloadToken({}), []);
+  const loadConfig = React.useCallback(async () => {
+    setStatus("loading");
+
+    try {
+      const [presetsList, fieldsList, metadata, messages] = await Promise.all([
+        api.getPresets(),
+        api.getFields(),
+        api.getMetadata(),
+        api.getConfigMessages(intl.locale),
+      ]);
+
+      if (!mountedRef.current) return; // if component was unmounted, don't set state
+
+      setConfig({
+        presets: new Map(
+          presetsList.filter(filterPointPreset).map(p => [p.id, p])
+        ),
+        fields: new Map(fieldsList.map(p => [p.id, p])),
+        metadata: metadata,
+        messages: messages,
+      });
+
+      setStatus("success");
+    } catch (err) {
+      bugsnag.notify(err, report => {
+        report.severity = "error";
+      });
+
+      log("Error loading presets and fields", err);
+
+      if (!mountedRef.current) return; // if component was unmounted, don't set state
+
+      setStatus("error");
+    }
+  }, [intl.locale]);
```

Comment:
There is a behaviour change and potential race condition introduced by this refactor. If the user changes the locale during the `await Promise.all()` call, then they could end up with the config showing in the previous locale. In the previous implementation, if `intl.locale` changed, then the config would be reloaded. It is highly unlikely this would ever happen, but I think best to avoid that bug.

I think this can be solved by changing how `api.getConfigMessages()` works, to return messages for all locales, so you can remove the `intl.locale` dependency for this useCallback. Then, in the `mergedMessages` memo below, get `config.messages[intl.locale]` and add `intl.local` to that useMemo() so that we get the right messages when the locale changes.

Reply:
Could we solve this by making the locale an argument for `loadConfig` and then calling it when the locale changes via a use effect?

```ts
const loadConfig = React.useCallback(async (locale: string) => {
  // ...
}

// Only *possible* issue is that it introduces a similar problem for the `replaceConfig` function
// but I think this would be less problematic/impossible to run into
// since it's called as a direct result of a user-initiated action
const replace = React.useCallback(
  async fileUri => {
    try {
      setStatus("loading");
      await api.replaceConfig(fileUri);
      await loadConfig(intl.locale);
    } catch (err) {
      log("Error replacing presets", err);
      setStatus("error");
    }
  },
  [intl.locale, loadConfig]
);

React.useEffect(() => {
  loadConfig(intl.locale)
}, [intl.locale])
```

Reply:
Implemented above change via https://github.com/digidem/mapeo-mobile/pull/730/commits/e360ef190730a9db3f6d4187406a5d9af44e963d